Prefer explicit securityContext to `privileged` flag

bleggett opened the pull request:
https://github.com/istio/istio/issues/49004


- Functionally speaking the required perms for sidecar repair and ambient are now pretty much 1:1.

- `CAP_SYS_ADMIN` is required for the proc mount scanning/opening netns FDs, at least, and I don't see a less broad capability that grants us that permission. We _might_ be able to skirt around this with a CSI driver to alter perms, and read-only mounts, but that's a lot of complexity and just moves the problem. We might could also solve it by not mounting `/proc` in the CNI agent, and conveying FDs from the CNI plugin itself, as has been batted around.

- With an eBPF impl we could probably drop CAP_NET_ADMIN and CAP_NET_RAW in favor of CAP_BPF, but practically speaking the operational risk is identical there.

bleggett committed:
Try this

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

bleggett committed:
Remove gen

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

linsun reviewed (commented):
LGTM
won't stamp approve to give others time to review.

howardjohn commented on manifests/charts/istio-cni/values.yaml:
```diff
@@ -40,11 +40,6 @@ defaults:
-    # Allow the istio-cni container to run in privileged mode, needed for some platforms (e.g. OpenShift) or features (repairPods)
-    # Note that even if this is false, the `istio-cni` container *requires* root privileges on the node to function,
-    # and setting this to false does not change that, nor will it run the container as non-root or make it "un-privileged".
-    privileged: true
```
I think openshift needs privilege for all operations, not just repair. I am not sure why exactly (it _might_ be the UDS log thing)

bleggett replied:
I wanna test this with openshift first - it should be fine with `securityContext`

It might need `privileged: true` in order to actually manually drop the `securityContext` privs down in the first place - in which case `privileged` should still be hardcoded.

I also wanna see if we can drop CAP_SYS_ADMIN entirely in favor of more granular capabilities, as it's quite overloaded.

linsun replied:
good point John. cc a few openshift folks to chime in @dgn @rcernich @jwendell

jacob-delgado replied:
There is platform value in the bae chart. Not sure if this could be expanded to the cni chart.

bleggett replied:
For now, I am hardcoding `privileged` to `true` if either repair or ambient are enabled.

It is redundant with CAP_SYS_ADMIN, and having both won't hurt, if one is required.


In general, it's better to rely on the mechanism of `drop: ALL` via `securityContext` and then explicitly adding back the CAP_XXX perms we actually require. 

Anything that doesn't trust us to run as `privileged` shouldn't/won't trust us to drop-all-add-SYS_ADMIN, and vice versa.

bleggett committed:
This is the best we can do for now

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

istio-testing commented:
@bleggett: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 279a0bcb44d86ecf26f2dcb67dff3d44214cdbb8 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49029/release-notes_istio/1752062761161986048) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@bleggett: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 279a0bcb44d86ecf26f2dcb67dff3d44214cdbb8 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49029/release-notes_istio/1752062761161986048) | true | `/test release-notes`
lint_istio | 279a0bcb44d86ecf26f2dcb67dff3d44214cdbb8 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49029/lint_istio/1752062760318930944) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

bleggett committed:
Relnotes

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

bleggett committed:
Remove `privileged: true` from templates

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

hzxuzhonghu commented on manifests/charts/istio-cni/templates/daemonset.yaml:
```diff
@@ -84,18 +84,25 @@ spec:
+            # privileged is redundant with CAP_SYS_ADMIN
+            # since it's redundant, hardcode it to `true`, then manually drop ALL + readd granular
+            # capabilities we actually require
+            privileged: true
```
why still need to set priviledged with the below capabilities added
 CAP_SYS_ADMIN
CAP_NET_ADMIN
CAP_NET_RAW

bleggett replied:
@hzxuzhonghu I don't have an OpenShift env yet to test with, so I can't rule out the possibility that OpenShift might require `privileged: true` to be set before it will allow a container to drop all capabilities and re-add granular ones.

Also, I think it's preferable in the short term to have both for compat reasons - if anyone has any, say, existing pod validation hooks that are checking for `privileged: true` but not checking `securityContext`, setting both means the pod would still get flagged as requiring elevated permissions.

Setting this to true here in this conditional basically means the templated pod manifest will be backwards compatible for all cases where `privileged` would have been set to true before.

hzxuzhonghu replied:
make sense

ymesika commented on operator/pkg/apis/istio/v1alpha1/values_types.proto:
```diff
@@ -82,8 +82,6 @@ message CNIConfig {
 
   Resources resources = 17;
 
-  google.protobuf.BoolValue privileged = 18;
```
Shouldn't we keep this field as reserved so that no one will re-use it in the future with different type?

bleggett replied:
I don't think so. Given how we actually use protos to "wrap" helm charts, just re-using wouldn't work.

But happy to be corrected.

howardjohn replied:
I think if we just drop it we will give an error to many users on upgrade.. I wonder if we should just keep it and ignore (we can add a warning it does nothing) rather than fail?

bleggett replied:
just marked is as deprecated in the proto, that should be good enough - helm itself doesn't care, our proto machinery is the only thing that would.

ymesika reviewed (commented):
LG
Left only one comment regarding the removed protobuf field.

howardjohn commented:
@jwendell @jacob-delgado are you able to review this for openshift?

dgn commented:
this does break on OpenShift when setting `cni.repair.enabled: false` (because privileges are missing to create the UDS listener):
```
2024-02-02T16:11:04.973546Z	info	Start a UDS server for CNI plugin logs
2024-02-02T16:11:04.973713Z	error	Failed to start up UDS Log Server: failed to create UDS listener: failed to listen on unix socket "/var/run/istio-cni/log.sock": listen unix /var/run/istio-cni/log.sock: bind: permission denied
Error: failed to create UDS listener: failed to listen on unix socket "/var/run/istio-cni/log.sock": listen unix /var/run/istio-cni/log.sock: bind: permission denied
```

bleggett commented:
> this does break on OpenShift when setting `cni.repair.enabled: false` (because privileges are missing to create the UDS listener):
> 
> ```
> 2024-02-02T16:11:04.973546Z	info	Start a UDS server for CNI plugin logs
> 2024-02-02T16:11:04.973713Z	error	Failed to start up UDS Log Server: failed to create UDS listener: failed to listen on unix socket "/var/run/istio-cni/log.sock": listen unix /var/run/istio-cni/log.sock: bind: permission denied
> Error: failed to create UDS listener: failed to listen on unix socket "/var/run/istio-cni/log.sock": listen unix /var/run/istio-cni/log.sock: bind: permission denied
> ```

Oh I missed that the UDS listener stuff was contingent on the `repair` flag :/ 

TYVM I'll take a look.

bleggett commented:
> this does break on OpenShift when setting `cni.repair.enabled: false` (because privileges are missing to create the UDS listener):
> 
> ```
> 2024-02-02T16:11:04.973546Z	info	Start a UDS server for CNI plugin logs
> 2024-02-02T16:11:04.973713Z	error	Failed to start up UDS Log Server: failed to create UDS listener: failed to listen on unix socket "/var/run/istio-cni/log.sock": listen unix /var/run/istio-cni/log.sock: bind: permission denied
> Error: failed to create UDS listener: failed to listen on unix socket "/var/run/istio-cni/log.sock": listen unix /var/run/istio-cni/log.sock: bind: permission denied
> ```

Sorry - when you say `this does break` do you mean this PR, or removing `privileged`/setting to false?

If the latter, then we're good, since this PR doesn't do that, and I guess that confirms if we need one we need the other.

bleggett commented:
> this does break on OpenShift when setting `cni.repair.enabled: false` (because privileges are missing to create the UDS listener):
> 
> ```
> 2024-02-02T16:11:04.973546Z	info	Start a UDS server for CNI plugin logs
> 2024-02-02T16:11:04.973713Z	error	Failed to start up UDS Log Server: failed to create UDS listener: failed to listen on unix socket "/var/run/istio-cni/log.sock": listen unix /var/run/istio-cni/log.sock: bind: permission denied
> Error: failed to create UDS listener: failed to listen on unix socket "/var/run/istio-cni/log.sock": listen unix /var/run/istio-cni/log.sock: bind: permission denied
> ```

kk - then that confirms we always need `privileged=true` for the `istio-cni` agent, period then. I bet even if we got past the log socket failure, with `privileged=false` the CNI plugin install would fail.

bleggett committed:
`privileged` is always `true` for `istio-cni`

Non-openshift envs just sometimes let you *be* privileged without actually
declaring it in the manifest.

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

bleggett commented:
This should be good to merge.

bleggett committed:
Fixup

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

bleggett commented:
cc / LFR @GregHanson / @Monkeyanator / @costinm / @hanxiaop / @jacob-delgado

istio-testing commented:
In response to a cherrypick label: #49029 failed to apply on top of branch "release-1.21":
```
Applying: Try this
Applying: Remove gen
Using index info to reconstruct a base tree...
M	operator/pkg/apis/istio/v1alpha1/values_types.pb.go
M	operator/pkg/apis/istio/v1alpha1/values_types.proto
Falling back to patching base and 3-way merge...
Auto-merging operator/pkg/apis/istio/v1alpha1/values_types.proto
CONFLICT (content): Merge conflict in operator/pkg/apis/istio/v1alpha1/values_types.proto
Auto-merging operator/pkg/apis/istio/v1alpha1/values_types.pb.go
CONFLICT (content): Merge conflict in operator/pkg/apis/istio/v1alpha1/values_types.pb.go
error: Failed to merge in the changes.
hint: Use 'git am --show-current-patch=diff' to see the failed patch
Patch failed at 0002 Remove gen
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".

```

istio-testing commented:
In response to a cherrypick label: new issue created for failed cherrypick: #49611

bleggett commented:
Given that I forgot about this one and 1.21 is now late in the game in terms of release, I am _not_ going to cherrypick this into 1.21, unless someone has an objection.
